src/js-impl/index.js

```diff
@@ -22,17 +23,15 @@ export const getCycle = d => {
 
     If `S` contains only a single city, `u`, then `len[S][u]` is `d[n - 1][u]`.
   */
-  const len = []
+  const len = new Float64Array(all * n)
```

Comment:
`S` can vary from 1 to 2<sup>*n* - 1</sup> - 1 inclusive. We leave space in the array for `S` = 0 as well although we don't use it. So in practice `S` ranges from 0 to 2<sup>*n* - 1</sup> - 1 inclusive, a range of 2<sup>*n* - 1</sup> possible values. Meanwhile, `u` ranges from 0 to *n* - 2 inclusive, a range of *n* - 1 possible values. So this should be `new Float64Array((1 << (n - 1)) * (n - 1))`. Compare with [what we do for the WebAssembly](https://github.com/qntm/held-karp/blob/1d2d1d6720d5db9baf3523c9a33f1068e62377bf/src/wasm-impl/index.js#L18-L19).

(`all * n` does work, as it is always larger than or equal to `(1 << (n - 1)) * (n - 1)`, but it allocates a little more memory than necessary.)

Likewise for `prev`, below.

This is, however, a great optimisation, can't believe I didn't think of it! 👍